File: replicator/go/internal/replicator.go
// Package replicator provides the core logic for replaying historical CSE FAST
// market data messages stored in a MongoDB collection.
//
// It includes a FAST stop-bit integer decoder, a binary payload decoder, and a
// Replicator type that queries MongoDB and processes each message in
// chronological order.
//
// Typical use:
//
//	r, err := replicator.New("mongodb://localhost:27017", "OmsTradingApi_CSE_FAST_DB", "FastIncomingMessages")
//	if err != nil {
//	    log.Fatal(err)
//	}
//	defer r.Close()
//	stats, err := r.Replay(ctx, start, end, replicator.Options{ProgressInterval: 100})
package internal

import (
	"context"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"
	"unicode"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// ---------------------------------------------------------------------------
// FAST template ID → name map (CSE FAST template set)
// ---------------------------------------------------------------------------

// TemplateNames maps FAST template IDs to human-readable message names.
var TemplateNames = map[int]string{
	1:  "Logon",
	2:  "Logout",
	3:  "Heartbeat",
	4:  "SecurityDefinitionRequest",
	7:  "MarketDataRequest",
	9:  "ApplicationMessageRequest",
	14: "SecurityDefinition",
	15: "SecurityStatus",
	18: "MDSnapshot",
	20: "MDIncrementalRefresh",
	24: "News",
	25: "MarketDataRequestReject",
	26: "BusinessMessageReject",
	27: "ApplicationMessageRequestAck",
	28: "ApplicationMessageReport",
}

// ---------------------------------------------------------------------------
// FAST binary decoder
// ---------------------------------------------------------------------------

// readStopBitUint reads a FAST stop-bit-encoded unsigned integer from data
// starting at offset. Returns (value, bytesRead). Returns (0, 0) on failure.
func readStopBitUint(data []byte, offset int) (int, int) {
	value := 0
	read := 0
	for i := offset; i < len(data) && read < 5; i++ {
		b := data[i]
		value = (value << 7) | int(b&0x7F)
		read++
		if b&0x80 != 0 { // stop bit set
			return value, read
		}
	}
	return 0, 0
}

// DecodedBinary holds the result of decoding a raw FAST binary payload.
type DecodedBinary struct {
	Hex          string
	TemplateID   int
	TemplateName string
	Strings      []string
	Integers     []int
}

// DecodeBinary decodes a raw FAST binary payload.
func DecodeBinary(data []byte, extraNames map[int]string) DecodedBinary {
	result := DecodedBinary{
		Hex:      strings.ToUpper(hex.EncodeToString(data)),
		Strings:  []string{},
		Integers: []int{},
	}

	if len(data) == 0 {
		return result
	}

	// Merged template name map
	tmap := make(map[int]string, len(TemplateNames)+len(extraNames))
	for k, v := range TemplateNames {
		tmap[k] = v
	}
	for k, v := range extraNames {
		tmap[k] = v
	}

	// First stop-bit uint is the FAST template ID.
	// read == 0 means the buffer is too short or the encoding is invalid
	// (not the same as a template ID whose value happens to be zero).
	tid, read := readStopBitUint(data, 0)
	if read > 0 {
		result.TemplateID = tid
		if name, ok := tmap[tid]; ok {
			result.TemplateName = name
		} else {
			result.TemplateName = fmt.Sprintf("Unknown(%d)", tid)
		}
	}
	// If read == 0 the payload is too short to contain a valid template header;
	// TemplateID stays at its zero value and TemplateName stays empty.

	// Extract printable ASCII runs (≥ 3 chars)
	var run []rune
	flush := func() {
		if len(run) >= 3 {
			result.Strings = append(result.Strings, string(run))
		}
		run = run[:0]
	}
	for _, b := range data {
		r := rune(b)
		if r >= 32 && r <= 126 && unicode.IsPrint(r) {
			run = append(run, r)
		} else {
			flush()
		}
	}
	flush()

	// Extract first ≤ 10 stop-bit integers
	pos := 0
	for i := 0; i < 10 && pos < len(data); i++ {
		v, rd := readStopBitUint(data, pos)
		if rd == 0 {
			break
		}
		result.Integers = append(result.Integers, v)
		pos += rd
	}

	return result
}

// ---------------------------------------------------------------------------
// MongoDB document model
// ---------------------------------------------------------------------------

// FastIncomingMessage mirrors the MongoDB FastIncomingMessages document schema.
type FastIncomingMessage struct {
	ID                 primitive.ObjectID `bson:"_id,omitempty"`
	Channel            string             `bson:"Channel"`
	PacketNum          int64              `bson:"PacketNum"`
	MsgName            string             `bson:"MsgName"`
	MsgText            string             `bson:"MsgText"`
	SendingDateTimeUtc time.Time          `bson:"SendingDateTimeUtc"`
	RawMsg             primitive.Binary   `bson:"RawMsg,omitempty"`
}

// ---------------------------------------------------------------------------
// Single-message processing
// ---------------------------------------------------------------------------

// ProcessResult holds the outcome of processing a single message.
type ProcessResult struct {
	Success bool
	Error   string
	Binary  *DecodedBinary
	Fields  map[string]interface{}
}

// ProcessMessage decodes one FastIncomingMessage.
func ProcessMessage(msg FastIncomingMessage, extraNames map[int]string) ProcessResult {
	// Guard: nothing to decode
	if msg.MsgText == "" && len(msg.RawMsg.Data) == 0 {
		return ProcessResult{
			Error: fmt.Sprintf("%s: empty message (no MsgText, no RawMsg)", msg.MsgName),
		}
	}

	if strings.HasPrefix(msg.MsgText, "RAW_HEX:") {
		return ProcessResult{
			Error: fmt.Sprintf("%s: RAW_HEX binary stub – not deserializable", msg.MsgName),
		}
	}
	if strings.HasPrefix(msg.MsgText, "RAW_EMPTY") {
		return ProcessResult{
			Error: fmt.Sprintf("%s: RAW_EMPTY packet", msg.MsgName),
		}
	}

	result := ProcessResult{Success: true}

	// 1. Decode binary payload when present
	rawBytes := msg.RawMsg.Data
	if len(rawBytes) == 0 && msg.MsgText != "" {
		// Try Base64-decoding MsgText in case it holds encoded binary
		if decoded, err := base64.StdEncoding.DecodeString(msg.MsgText); err == nil {
			rawBytes = decoded
		}
	}
	if len(rawBytes) > 0 {
		b := DecodeBinary(rawBytes, extraNames)
		result.Binary = &b
	}

	// 2. Parse structured JSON fields from MsgText
	if msg.MsgText != "" {
		var fields map[string]interface{}
		if err := json.Unmarshal([]byte(msg.MsgText), &fields); err == nil {
			result.Fields = fields
		}
	}

	// At least one decode path must have produced output
	if result.Binary == nil && result.Fields == nil {
		return ProcessResult{
			Error: fmt.Sprintf("could not decode %s", msg.MsgName),
		}
	}

	return result
}

// ---------------------------------------------------------------------------
// Replicator
// ---------------------------------------------------------------------------

// Options controls replay behaviour.
type Options struct {
	// ProgressInterval prints a status line every N successfully processed messages.
	// Defaults to 100.
	ProgressInterval int
	// Verbose prints per-message details for MDIncrementalRefresh messages.
	Verbose bool
	// ExtraTemplateNames extends TemplateNames with additional ID→name mappings.
	ExtraTemplateNames map[int]string
}

// Stats summarises a completed replay.
type Stats struct {
	Total        int
	Processed    int
	Errors       int
	ErrorSummary map[string]int
}

// Replicator connects to MongoDB and replays FastIncomingMessage documents.
type Replicator struct {
	mongoURI       string
	dbName         string
	collectionName string
	client         *mongo.Client
	collection     *mongo.Collection
}

// New creates a new Replicator and opens the MongoDB connection.
// Call Close() when done.
func New(mongoURI, dbName, collectionName string) (*Replicator, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	clientOpts := options.Client().ApplyURI(mongoURI).
		SetServerSelectionTimeout(5 * time.Second)

	client, err := mongo.Connect(ctx, clientOpts)
	if err != nil {
		return nil, fmt.Errorf("connect: %w", err)
	}
	// Probe connectivity
	if err := client.Ping(ctx, nil); err != nil {
		_ = client.Disconnect(context.Background())
		return nil, fmt.Errorf("ping: %w", err)
	}

	col := client.Database(dbName).Collection(collectionName)
	return &Replicator{
		mongoURI:       mongoURI,
		dbName:         dbName,
		collectionName: collectionName,
		client:         client,
		collection:     col,
	}, nil
}

// Close disconnects from MongoDB.
func (r *Replicator) Close() {
	if r.client != nil {
		_ = r.client.Disconnect(context.Background())
	}
}

// StatusInfo holds collection statistics.
type StatusInfo struct {
	Total        int64
	Earliest     time.Time
	Latest       time.Time
	MessageTypes []MessageTypeCount
}

// MessageTypeCount holds a (name, count) pair.
type MessageTypeCount struct {
	Name  string
	Count int64
}

// Status returns statistics about the collection.
func (r *Replicator) Status(ctx context.Context) (StatusInfo, error) {
	var info StatusInfo

	total, err := r.collection.CountDocuments(ctx, bson.D{})
	if err != nil {
		return info, err
	}
	info.Total = total

	if total == 0 {
		return info, nil
	}

	// Earliest message
	var earliest FastIncomingMessage
	opts := options.FindOne().SetSort(bson.D{{Key: "SendingDateTimeUtc", Value: 1}})
	if err := r.collection.FindOne(ctx, bson.D{}, opts).Decode(&earliest); err == nil {
		info.Earliest = earliest.SendingDateTimeUtc
	}

	// Latest message
	var latest FastIncomingMessage
	opts = options.FindOne().SetSort(bson.D{{Key: "SendingDateTimeUtc", Value: -1}})
	if err := r.collection.FindOne(ctx, bson.D{}, opts).Decode(&latest); err == nil {
		info.Latest = latest.SendingDateTimeUtc
	}

	// Per-type counts via aggregation
	pipeline := mongo.Pipeline{
		{{Key: "$group", Value: bson.D{
			{Key: "_id", Value: "$MsgName"},
			{Key: "count", Value: bson.D{{Key: "$sum", Value: 1}}},
		}}},
		{{Key: "$sort", Value: bson.D{{Key: "_id", Value: 1}}}},
	}
	cursor, err := r.collection.Aggregate(ctx, pipeline)
	if err != nil {
		return info, err
	}
	defer cursor.Close(ctx)

	for cursor.Next(ctx) {
		var row struct {
			Name  string `bson:"_id"`
			Count int64  `bson:"count"`
		}
		if err := cursor.Decode(&row); err == nil {
			info.MessageTypes = append(info.MessageTypes, MessageTypeCount{
				Name:  row.Name,
				Count: row.Count,
			})
		}
	}
	return info, cursor.Err()
}

// PrintStatus prints a human-readable status to stdout.
func (r *Replicator) PrintStatus(ctx context.Context) error {
	info, err := r.Status(ctx)
	if err != nil {
		return err
	}
	fmt.Printf("  Total messages : %d\n", info.Total)
	if info.Total == 0 {
		fmt.Println("  WARNING: Collection is empty – no messages to replay.")
		return nil
	}
	if !info.Earliest.IsZero() && !info.Latest.IsZero() {
		fmt.Printf("  Date range     : %s UTC  →  %s UTC\n",
			info.Earliest.UTC().Format("2006-01-02 15:04:05"),
			info.Latest.UTC().Format("2006-01-02 15:04:05"))
	}
	fmt.Printf("  Message types  : %d\n", len(info.MessageTypes))
	for _, mt := range info.MessageTypes {
		fmt.Printf("    %-40s  %8d\n", mt.Name, mt.Count)
	}
	return nil
}

// GetMessages returns all messages in the half-open interval [start, end).
func (r *Replicator) GetMessages(ctx context.Context, start, end time.Time) ([]FastIncomingMessage, error) {
	filter := bson.D{
		{Key: "SendingDateTimeUtc", Value: bson.D{
			{Key: "$gte", Value: start},
			{Key: "$lt", Value: end},
		}},
	}
	opts := options.Find().SetSort(bson.D{{Key: "SendingDateTimeUtc", Value: 1}})
	cursor, err := r.collection.Find(ctx, filter, opts)
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var msgs []FastIncomingMessage
	if err := cursor.All(ctx, &msgs); err != nil {
		return nil, err
	}
	return msgs, nil
}

// Replay processes all messages in [start, end) and returns summary statistics.
func (r *Replicator) Replay(ctx context.Context, start, end time.Time, opts Options) (Stats, error) {
	if opts.ProgressInterval <= 0 {
		opts.ProgressInterval = 100
	}

	messages, err := r.GetMessages(ctx, start, end)
	if err != nil {
		return Stats{}, err
	}

	total := len(messages)
	processed := 0
	errors := 0
	errorSummary := make(map[string]int)
	errorExamples := make(map[string]string)

	fmt.Printf("Found %d messages to replay\n\n", total)

	for _, msg := range messages {
		result := ProcessMessage(msg, opts.ExtraTemplateNames)
		tsStr := msg.SendingDateTimeUtc.UTC().Format("2006-01-02 15:04:05.000")

		if result.Success {
			processed++

			if opts.Verbose && msg.MsgName == "MDIncrementalRefresh" {
				entries := 0
				if result.Fields != nil {
					if incRef, ok := result.Fields["IncRefMDEntries"].(map[string]interface{}); ok {
						if items, ok := incRef["Items"].([]interface{}); ok {
							entries = len(items)
						}
					}
				}
				fmt.Printf("  ✓ MDIncrementalRefresh  %s  ch=%s  entries=%d\n",
					tsStr, msg.Channel, entries)
			}

			if processed%opts.ProgressInterval == 0 {
				fmt.Printf("  Processed %d messages...\n", processed)
			}
		} else {
			errors++
			key := result.Error
			if key == "" {
				key = "Unknown error"
			}
			errorSummary[key]++
			if _, seen := errorExamples[key]; !seen {
				errorExamples[key] = fmt.Sprintf("%s  ch=%s", tsStr, msg.Channel)
			}
		}
	}

	fmt.Println()
	fmt.Println("Summary:")
	fmt.Printf("  Total messages       : %d\n", total)
	fmt.Printf("  Successfully decoded : %d\n", processed)
	fmt.Printf("  Errors               : %d\n", errors)

	if len(errorSummary) > 0 {
		fmt.Println()
		fmt.Println("Error Breakdown:")

		// Sort errors by count descending
		type kv struct {
			Key   string
			Count int
		}
		var sorted []kv
		for k, v := range errorSummary {
			sorted = append(sorted, kv{k, v})
		}
		sort.Slice(sorted, func(i, j int) bool { return sorted[i].Count > sorted[j].Count })

		for _, item := range sorted {
			fmt.Printf("  %6dx  %s\n", item.Count, item.Key)
			if ex, ok := errorExamples[item.Key]; ok {
				fmt.Printf("          first: %s\n", ex)
			}
		}
	}

	return Stats{
		Total:        total,
		Processed:    processed,
		Errors:       errors,
		ErrorSummary: errorSummary,
	}, nil
}

File: replicator/replicator.py
#!/usr/bin/env python3
"""
replicator.py — CSE FAST Market Data Replay Tool (Python version)

Replays historical FAST market data messages from a MongoDB collection,
decoding binary payloads and structured JSON fields for analysis or
testing purposes.

This file has been extended to optionally PUBLISH RawMsg payloads to a
multicast group/port (e.g. 239.100.140.50:7540) during replay.

Requirements:
    pip install pymongo

Usage (library):
    from replicator import FastReplicator
    r = FastReplicator("mongodb://localhost:27017", "OmsTradingApi_CSE_FAST_DB")
    r.connect()
    r.replay(start, end)
    r.close()
"""

from __future__ import annotations

import base64
import json
import socket
import struct
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, List, Optional, Tuple, Any


# ---------------------------------------------------------------------------
# Multicast config helpers
# ---------------------------------------------------------------------------

DEFAULT_MCAST_IP = "239.100.140.50"
DEFAULT_MCAST_PORT = 7540
DEFAULT_MCAST_TTL = 1


def _load_appsettings(appsettings_path: Path) -> Dict[str, Any]:
    """
    Best-effort load of appsettings.json-like file.
    Returns {} if not found or invalid.
    """
    try:
        if not appsettings_path.exists():
            return {}
        return json.loads(appsettings_path.read_text(encoding="utf-8"))
    except Exception:
        return {}


def _get_mcast_settings(appsettings_path: Optional[Path] = None) -> Tuple[str, int, int]:
    """
    Returns (ip, port, ttl) from appsettings.json (if present), otherwise defaults.
    """
    if appsettings_path is None:
        # replicator.py lives in replicator/
        appsettings_path = Path(__file__).resolve().parent / "appsettings.json"

    cfg = _load_appsettings(appsettings_path)

    ip = cfg.get("MultiCastConnectionIP", DEFAULT_MCAST_IP)

    port_raw = cfg.get("MultiCastConnectionPort", DEFAULT_MCAST_PORT)
    try:
        port = int(port_raw)
    except Exception:
        port = DEFAULT_MCAST_PORT

    # Optional support if you later add it
    ttl_raw = cfg.get("MultiCastTTL", DEFAULT_MCAST_TTL)
    try:
        ttl = int(ttl_raw)
    except Exception:
        ttl = DEFAULT_MCAST_TTL

    return str(ip), port, ttl


class MulticastPublisher:
    """
    UDP multicast publisher (sender).
    """

    def __init__(self, group_ip: str, port: int, ttl: int = 1) -> None:
        self.group_ip = group_ip
        self.port = port
        self.ttl = ttl

        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
        # TTL as a single signed byte
        self.sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, struct.pack("b", int(ttl)))

    def send(self, payload: bytes) -> None:
        self.sock.sendto(payload, (self.group_ip, self.port))

    def close(self) -> None:
        try:
            self.sock.close()
        except Exception:
            pass


def _rawmsg_to_bytes(raw_msg: Any) -> Optional[bytes]:
    """
    Convert MongoDB RawMsg field (bytes or base64 string) to bytes.
    """
    if raw_msg is None:
        return None

    if isinstance(raw_msg, bytes):
        return raw_msg

    if isinstance(raw_msg, str):
        # stored base64 in Mongo?
        try:
            return base64.b64decode(raw_msg)
        except Exception:
            return None

    return None


# ---------------------------------------------------------------------------
# FAST binary decoder (stop-bit encoding)
# ---------------------------------------------------------------------------

#: Template IDs → human-readable names (CSE FAST template set)
TEMPLATE_NAMES: Dict[int, str] = {
    1:  "Logon",
    2:  "Logout",
    3:  "Heartbeat",
    4:  "SecurityDefinitionRequest",
    7:  "MarketDataRequest",
    9:  "ApplicationMessageRequest",
    14: "SecurityDefinition",
    15: "SecurityStatus",
    18: "MDSnapshot",
    20: "MDIncrementalRefresh",
    24: "News",
    25: "MarketDataRequestReject",
    26: "BusinessMessageReject",
    27: "ApplicationMessageRequestAck",
    28: "ApplicationMessageReport",
}


def _read_stop_bit_uint(data: bytes, offset: int) -> Tuple[int, int]:
    """Read a FAST stop-bit-encoded unsigned integer from *data* at *offset*.

    Returns ``(value, bytes_read)``.  Returns ``(0, 0)`` when no valid
    integer can be decoded (e.g. offset beyond end of buffer).
    """
    value = 0
    read = 0
    for i in range(offset, min(offset + 5, len(data))):
        b = data[i]
        value = (value << 7) | (b & 0x7F)
        read += 1
        if b & 0x80:           # stop bit set → last byte
            return value, read
    return 0, 0


def decode_binary(data: bytes, template_map: Optional[Dict[int, str]] = None) -> Dict:
    """Decode a raw FAST binary payload.

    Returns a dict with keys:

    * ``hex``           – upper-case hex string of the raw bytes
    * ``template_id``   – integer template ID (or ``None``)
    * ``template_name`` – resolved name from *template_map* (or ``None``)
    * ``strings``       – list of printable ASCII runs ≥ 3 chars
    * ``integers``      – first ≤ 10 stop-bit integers decoded from head
    """
    tmap = {**TEMPLATE_NAMES, **(template_map or {})}
    result: Dict = {
        "hex": data.hex().upper(),
        "template_id": None,
        "template_name": None,
        "strings": [],
        "integers": [],
    }

    if not data:
        return result

    # First stop-bit uint is the FAST template ID
    tid, read = _read_stop_bit_uint(data, 0)
    if read > 0:
        result["template_id"] = tid
        result["template_name"] = tmap.get(tid, f"Unknown({tid})")

    # Extract printable ASCII runs
    run: List[str] = []
    for b in data:
        if 32 <= b <= 126:
            run.append(chr(b))
        else:
            if len(run) >= 3:
                result["strings"].append("".join(run))
            run = []
    if len(run) >= 3:
        result["strings"].append("".join(run))

    # Extract first ≤ 10 stop-bit integers
    pos = 0
    for _ in range(10):
        if pos >= len(data):
            break
        val, rd = _read_stop_bit_uint(data, pos)
        if rd == 0:
            break
        result["integers"].append(val)
        pos += rd

    return result


# ---------------------------------------------------------------------------
# Single-message processing
# ---------------------------------------------------------------------------

def process_message(msg: Dict, template_map: Optional[Dict[int, str]] = None) -> Dict:
    """Process one ``FastIncomingMessage`` document from MongoDB.

    Returns a dict with:

    * ``success`` (bool)
    * ``error``   (str | None) – populated only on failure
    * ``binary``  (dict | None) – result of :func:`decode_binary`
    * ``fields``  (dict | None) – parsed ``MsgText`` JSON
    """
    msg_name = msg.get("MsgName") or "Unknown"
    msg_text = msg.get("MsgText") or ""
    raw_msg  = msg.get("RawMsg")

    # Guard: nothing to decode
    if not msg_text and not raw_msg:
        return {"success": False, "error": f"{msg_name}: empty message (no MsgText, no RawMsg)"}

    if isinstance(msg_text, str):
        if msg_text.startswith("RAW_HEX:"):
            return {"success": False, "error": f"{msg_name}: RAW_HEX binary stub – not deserializable"}
        if msg_text.startswith("RAW_EMPTY"):
            return {"success": False, "error": f"{msg_name}: RAW_EMPTY packet"}

    result: Dict = {"success": True, "error": None, "binary": None, "fields": None}

    # 1. Decode binary payload when present
    if raw_msg:
        raw_bytes = _rawmsg_to_bytes(raw_msg)
        if raw_bytes:
            try:
                result["binary"] = decode_binary(raw_bytes, template_map)
            except Exception:
                pass  # non-fatal, fall through to JSON path

    # 2. Parse structured JSON fields from MsgText
    if msg_text:
        try:
            result["fields"] = json.loads(msg_text)
        except Exception:
            pass

    # At least one decode path must have succeeded
    if result["binary"] is None and result["fields"] is None:
        return {"success": False, "error": f"Could not decode {msg_name}"}

    return result


# ---------------------------------------------------------------------------
# Replicator class
# ---------------------------------------------------------------------------

class FastReplicator:
    """Replays historical FAST market data from a MongoDB collection.

    Parameters
    ----------
    mongo_uri:
        Full MongoDB connection URI, e.g. ``"mongodb://localhost:27017"``.
    db_name:
        Name of the database, e.g. ``"OmsTradingApi_CSE_FAST_DB"``.
    collection_name:
        MongoDB collection holding ``FastIncomingMessage`` documents.
    template_map:
        Optional dict of ``{template_id: name}`` to override/extend the
        built-in :data:`TEMPLATE_NAMES`.

    Multicast publishing
    --------------------
    If publish_multicast=True, RawMsg (bytes/base64) will be sent via UDP
    to MultiCastConnectionIP:MultiCastConnectionPort from appsettings.json.
    """

    def __init__(
        self,
        mongo_uri: str = "mongodb://localhost:27017",
        db_name: str = "OmsTradingApi_CSE_FAST_DB",
        collection_name: str = "FastIncomingMessages",
        template_map: Optional[Dict[int, str]] = None,
    ) -> None:
        self.mongo_uri = mongo_uri
        self.db_name = db_name
        self.collection_name = collection_name
        self.template_map = template_map or {}
        self._client = None
        self._collection = None

    # ------------------------------------------------------------------
    # Connection management
    # ------------------------------------------------------------------

    def connect(self, timeout_ms: int = 5000) -> None:
        """Open a connection to MongoDB and verify it is reachable."""
        try:
            from pymongo import MongoClient  # type: ignore[import]
        except ImportError as exc:
            raise ImportError(
                "pymongo is required: pip install pymongo"
            ) from exc

        self._client = MongoClient(self.mongo_uri, serverSelectionTimeoutMS=timeout_ms)
        # Probe connectivity (raises if server is unreachable)
        self._client.server_info()
        db = self._client[self.db_name]
        self._collection = db[self.collection_name]

    def close(self) -> None:
        """Close the MongoDB connection."""
        if self._client is not None:
            self._client.close()
            self._client = None
            self._collection = None

    # ------------------------------------------------------------------
    # Database statistics
    # ------------------------------------------------------------------

    def get_status(self) -> Dict:
        """Return a status dictionary describing the collection contents."""
        self._require_connection()
        col = self._collection

        total = col.count_documents({})
        status: Dict = {"total": total, "earliest": None, "latest": None, "message_types": []}

        if total == 0:
            return status

        earliest = col.find_one({}, sort=[("SendingDateTimeUtc", 1)])
        latest   = col.find_one({}, sort=[("SendingDateTimeUtc", -1)])
        status["earliest"] = earliest.get("SendingDateTimeUtc") if earliest else None
        status["latest"]   = latest.get("SendingDateTimeUtc")   if latest   else None

        pipeline = [
            {"$group": {"_id": "$MsgName", "count": {"$sum": 1}}},
            {"$sort":  {"_id": 1}},
        ]
        status["message_types"] = list(col.aggregate(pipeline))
        return status

    def print_status(self) -> None:
        """Print a human-readable database status to stdout."""
        s = self.get_status()
        print(f"  Total messages : {s['total']:,}")
        if s["total"] == 0:
            print("  WARNING: Collection is empty – no messages to replay.")
            return
        if s["earliest"] and s["latest"]:
            fmt = "%Y-%m-%d %H:%M:%S"
            print(f"  Date range     : {s['earliest'].strftime(fmt)} UTC  →  {s['latest'].strftime(fmt)} UTC")
        print(f"  Message types  : {len(s['message_types'])}")
        for mt in s["message_types"]:
            name  = mt.get("_id") or "Unknown"
            count = mt.get("count", 0)
            print(f"    {name:<40}  {count:>8,}")

    # ------------------------------------------------------------------
    # Query helpers
    # ------------------------------------------------------------------

    def get_messages(self, start: datetime, end: datetime) -> List[Dict]:
        """Return all messages whose ``SendingDateTimeUtc`` is in ``[start, end)``."""
        self._require_connection()
        query = {"SendingDateTimeUtc": {"$gte": start, "$lt": end}}
        return list(self._collection.find(query).sort("SendingDateTimeUtc", 1))

    # ------------------------------------------------------------------
    # Replay
    # ------------------------------------------------------------------

    def replay(
        self,
        start: datetime,
        end: datetime,
        progress_interval: int = 100,
        verbose: bool = False,
        publish_multicast: bool = True,
    ) -> Dict:
        """Replay all messages in the date window ``[start, end)``.

        Parameters
        ----------
        start / end:
            UTC-aware or naïve ``datetime`` objects defining the window.
        progress_interval:
            Print a progress line every *N* successfully processed messages.
        verbose:
            When ``True`` print per-message details for MDIncrementalRefresh.
        publish_multicast:
            When True, send RawMsg bytes to the multicast group/port.

        Returns
        -------
        dict
            ``{"total": int, "processed": int, "errors": int,
               "published": int, "publish_errors": int,
               "error_summary": {str: int}}``
        """
        messages  = self.get_messages(start, end)
        total     = len(messages)
        processed = 0
        errors    = 0
        published = 0
        publish_errors = 0
        error_summary: Dict[str, int] = {}
        error_examples: Dict[str, str] = {}

        pub: Optional[MulticastPublisher] = None
        if publish_multicast:
            ip, port, ttl = _get_mcast_settings()
            pub = MulticastPublisher(ip, port, ttl)
            print(f"Multicast publish: {ip}:{port} (ttl={ttl})")
            print()

        print(f"Found {total:,} messages to replay")
        print()

        try:
            for msg in messages:
                result = process_message(msg, self.template_map)
                ts_str = ""
                ts_val = msg.get("SendingDateTimeUtc")
                if isinstance(ts_val, datetime):
                    ts_str = ts_val.strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]

                if result["success"]:
                    processed += 1

                    # Publish RawMsg (FAST binary) via multicast
                    if pub is not None:
                        raw_bytes = _rawmsg_to_bytes(msg.get("RawMsg"))
                        if raw_bytes:
                            try:
                                pub.send(raw_bytes)
                                published += 1
                            except Exception:
                                publish_errors += 1

                    if verbose and msg.get("MsgName") == "MDIncrementalRefresh":
                        flds  = result.get("fields") or {}
                        items = flds.get("IncRefMDEntries", {}).get("Items") if isinstance(flds, dict) else None
                        count = len(items) if isinstance(items, list) else 0
                        print(f"  ✓ MDIncrementalRefresh  {ts_str}  ch={msg.get('Channel')}  entries={count}")

                    if processed % progress_interval == 0:
                        print(f"  Processed {processed:,} messages...")
                else:
                    errors += 1
                    key = result.get("error") or "Unknown error"
                    error_summary[key] = error_summary.get(key, 0) + 1
                    if key not in error_examples:
                        ch = msg.get("Channel", "")
                        error_examples[key] = f"{ts_str}  ch={ch}"
        finally:
            if pub is not None:
                pub.close()

        print()
        print("Summary:")
        print(f"  Total messages       : {total:,}")
        print(f"  Successfully decoded : {processed:,}")
        print(f"  Errors               : {errors:,}")
        if publish_multicast:
            print(f"  Published (RawMsg)   : {published:,}")
            print(f"  Publish errors       : {publish_errors:,}")

        if error_summary:
            print()
            print("Error Breakdown:")
            for key, cnt in sorted(error_summary.items(), key=lambda x: -x[1]):
                print(f"  {cnt:6}x  {key}")
                if key in error_examples:
                    print(f"          first: {error_examples[key]}")

        return {
            "total":          total,
            "processed":      processed,
            "errors":         errors,
            "published":      published,
            "publish_errors": publish_errors,
            "error_summary":  error_summary,
        }

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _require_connection(self) -> None:
        if self._collection is None:
            raise RuntimeError("Not connected. Call connect() first.")
